### PR TITLE
Always post success/failure comment

### PR DESCRIPTION
**Pull request by luhenry**
Fixes https://github.com/luhenry/monors/issues/12

### CODE REVIEW

**Comment by akoeplinger** [comment]
As discussed on Slack I'm not a fan of doing this on all PRs given how spammy it is. It'll also use up even more precious GitHub token usages. I'm fine with making this an optional feature like `@monojenkins notify on failures`.

**Comment by luhenry** [comment]
It would be similar to what's already done on XI (ex: https://github.com/xamarin/xamarin-macios/pull/4996#issuecomment-430363140), and it's very helpful when reviewing the PR since you can more easily figure out if failures are flaky or related (if it fails 3 times in a row, that's a strong signal that something is broken and not just flaky).

**Comment by akoeplinger** [comment]
yeah and I hate the feature over there since all you see are those build result comments :D

**Comment by akoeplinger** [comment]
you can get the same flaky results traces from the Slack bot

**Comment by luhenry** [comment]
@akoeplinger the problem the Slack bot is addressing is not the same as the problem the proposal is trying to address: this isn't a feature for the author of the PR (which the Slack bot is), it's a tool for reviewers as well to understand better what's the failure history of this PR. Right now, if you do `@monojenkins build failed`, you have no clue what failures happened and the only way to know that is to go over all Jenkins jobs and try to find the last failure for the said PR (ie it's not doable on a regular basis).

@lewurm what's your take on that?

**Comment by lewurm** [comment]
I like this idea 👍  A few comments:

* I would like to see successful configurations included as well, wrapped with [`<details>`](https://gist.github.com/ericclemmons/b146fe5da72ca1f706b2ef72a20ac39d), so it's less noisy in the github PR UI. This might not be very useful right now, but once #10 is implemented, it can be useful to have that data available.

* Slack bot is not enough, as often this kind of information is for reviewers interesting as well. FWIW, I like the bot that the Xamarin.iOS team uses.

* > It'll also use up even more precious GitHub token usages.

    That's an implementation detail. I know it's going to be a problem, but we shouldn't avoid adding useful features because of this.

File: monors.py [hunk 1]
```diff
@@ -217,6 +217,56 @@ def load_statuses (self):
...
+                    logging.info (message)
+
+                    if status_text != "success":
+                        comment += message + "\n"
+
+            last_comment = None
+            for c in self.comments:
+                if c.login != self.cfg ["user"].encode ("utf8"):
+                    continue
+                if not c.body.startswith ("Required status checks are not successful:".encode ("utf8")) and not c.body.startswith ("Required status checks are successful.".encode ("utf8"))):
```
Comment (lewurm) [inline monors.py:249]: I guess we can simplify this to check
```suggestion
                if not c.body.startswith ("Required status checks are ".encode ("utf8")):
```

**Comment by akoeplinger** [comment]
>> It'll also use up even more precious GitHub token usages.

> That's an implementation detail. I know it's going to be a problem, but we shouldn't avoid adding useful features because of this.

It's a real problem, we're already hitting 0 remaining tokens at least once a week and then everything starts to break down...

**Comment by akoeplinger** [comment]
What if we included the build history in the PR description and the bot would just update/append to it? 
Cause it sounds like you're more interested in the history rather than getting the notification right?

**Comment by luhenry** [comment]
@akoeplinger the signal I hope we get out of that - and that we lost - is: when doing `@monojenkins build failed`, which builds are failing consistently. As a reviewer this is quite important since it will give me a strong signal to look at these specific builds' failures and try to understand why it happened. By aggregating that in the PR's description, you lose that connection `@monojenkins build failed` <-> `The failed builds are ...`.

The notification is actually a nice side effect since I am not iterating over *all* PRs *everyday* so getting a notification when builds are finished would signal me that I should look at a PR again, since there may be build failures/success that I was waiting for.

As for the GitHub token issues, it's been discussed yesterday with Julia, and it seems the .NET/Roslyn team had already run into the issue and their solution was a workaround with GitHub. More details in PM if you want.

**Comment by akoeplinger** [comment]
As discussed on Slack we'll postpone this until we have time to discuss with the whole team in December.

**Comment by luhenry** [comment]
@akoeplinger @lewurm other possible approach: we post the list of failed builds on the PR as part of the `@monojenkins build failed` command. Right before re-triggering the failed builds, we post a comment with the list of failed builds with their URLs. This would give us visibility in what exactly fails, while still minimising the number of comments. We would lose the side effect of the notification, but that's something I can live with.

This would mean closing this PR and submitting another one on the appropriate bot that listen to the `@monojenkins build failed` trigger. @akoeplinger could you please share where I can find this bot?

**Comment by akoeplinger** [comment]
@luhenry that sounds like a better approach.

We could even make the monojenkins bot _edit_ the comment that did `@monojenkins build failed` and append the list of things it's about to retry, that way we don't even need a separate comment. It should work because monojenkins is admin on the repo so it can edit any comment.

The bot which handles this is in https://github.com/xamarin/github-backport-bot/blob/master/mono-github-trigger/CommentTrigger.cs

**Comment by luhenry** [comment]
@akoeplinger I am not a fan of editing the existing comment, for linearity of history sake, but I'll look at submitting a PR against the existing code.

**Comment by akoeplinger** [comment]
Sure, we can also discuss it a bit more on Slack if you want :)

**Comment by luhenry** [comment]
Going to implement that feature in https://github.com/xamarin/mono-github-trigger/ instead.

**Pull request closed without merging**